README.md
# Welcome to Pogo

### Pogo is a GUI application for generating Tango projects.
 See Wiki for more insformation

#### Pogo-9.7.0 needs to be compiled and executed with JAVA 11 or higher.
[![Download](https://api.bintray.com/packages/tango-controls/maven/Pogo/images/download.svg) ](https://bintray.com/tango-controls/maven/Pogo/_latestVersion) Latest release

#### The latest release compatible with java 8 is  Pogo-9.6.31
[Download](https://bintray.com/tango-controls/maven/download_file?file_path=org%2Ftango%2Ftools%2Fpogo%2Fgui%2FPogo%2F9.6.31%2FPogo-9.6.31.jar)
Latest release **java 8** compatible

[![Docs](https://img.shields.io/badge/Latest-Docs-orange.svg)](https://tango-controls.github.io/pogo/)




fr.esrf.tango.pogo.parent/CHANGELOG.md
# Changelog


#### Pogo-9.7.2 - 17/06/2020:
    Fix problem on displayed release number

#### Pogo-9.7.0 - 16/06/2020:
    Needs to be compiled and executed with JAVA 11 or higher (JAVA 8 compatibility lost)

#### Pogo-9.6.31 - 01/04/2020:
    Last release compiled and executed with JAVA 8
    Pull request #94 integration

#### Pogo-9.6.28 - 06/02/2020:
    Pull requests #87 and #91 integration

#### Pogo-9.6.27 - 16/01/2020:
    Pull request #81 integration

#### Pogo-9.6.26 - 06/11/2019:
    Update base classe inheritance from PyTango.Device4Impl to PyTango.LatestDeviceImpl ( Pull request #71)

#### Pogo-9.6.25 - 24/10/2019:
    Fix a problem when conning property Class/Device
    Pull request #72 from SKA

#### Pogo-9.6.24 - 22/10/2019:
    Forbids Copy/Clone and Remove for State and status
    Added '_' character in autorizedChars list in EnumDialog.java code.
    Changed imports from PyTango to tango.

#### Pogo-9.6.21 - 27/05/2019:
    Remove extra ";" warning (#60)

#### Pogo-9.6.20 - 21/05/2019:
    Improve additional files management in CMakelist.txt 

#### Pogo-9.6.19 - 10/05/2019:
    Henance exception management if a non xmi file is loaded

#### Pogo-9.6.18 - 15/11/2018:
    Fix a bug in copy/paste object.

#### Pogo-9.6.17 - 09/11/2018:
    Fix a bug in dimX/dimY on java image attributes.

#### Pogo-9.6.16 - 03/10/2018:
    When generate html doc, if NO_PDF=true, does not generate (Soleil request)

#### Pogo-9.6.15 - 09/07/2018:
    Fix a problem LowerCase/UpperCase.

#### Pogo-9.6.14 - 09/07/2018:
    Fix a bug in relative path for inheritance (PR code)

#### Pogo-9.6.13 - 19/06/2018:
    Fix bug in relative path for multiple classes.
    It is now possible to load a server even a class has not be found.

#### Pogo-9.6.12 - 11/06/2018:
    Add protected regions in dynamic attribute and command classes

#### Pogo-9.6.11 - 04/06/2018:
    Fix a problem on dynamic attributes when no database.

#### Pogo-9.6.9 - 03/04/2018:
    Fix aproblem in CMakelist.txt file.
    Add new feature to get enum labels from file.

#### Pogo-9.6.6 - 12/03/2018:
    Fix aproblem in Enum labels for java classes.

#### Pogo-9.6.5 - 23/02/2018:
    Fix a bug on heritance path.
    Generate Makefile or CMakeList.txt from commande line.

#### Pogo-9.6.4 - 12/01/2018:
    Change default (1000) and minimum (5) values for polling
    Remove references to Pogo-6 (too old)

#### Pogo-9.6.3 - 09/10/2017:
    Fix a NPE in AttributeDialog.java

#### Pogo-9.6.2 - 06/10/2017:
    Allow to generate ENUM attribute as spectrum and image in C++.

#### Pogo-9.6.1 - 25/09/2017:
    Fix a NPE in PyHL new class creation.

#### Pogo-9.6.0 - 15/09/2017:
    PyHL package is now memorized for next reload/generate action.
    Fix bug in Pogo command line.

#### Pogo-9.5.2 - 12/08/2017:
    CMakeList added for widows (thanks to Sebastien)

#### Pogo-9.5.1 - 10/08/2017:
    Add release notes and remove unused files.

#### Pogo-9.5.0 - 09/08/2017:
    Mavenized project (thanks to Igor)

#### Pogo-9.4.6a - 19/04/2017:
    Cosmetic changes

#### Pogo-9.4.6 - 13/01/2017:
    Add a pdf file in doc_html on linux (if wkhtmltopdf found in path)
    Fix a bug in html protected regions.
    Fix a bug in key word management.
    Add polling for State/Status in java.

#### Pogo-9.4.5 - 08/12/2016:
    A pom.xml file could be generated for java server.
    Path must be in respect with maven structure (--/--/src/main/java)

#### Pogo-9.4.4 - 06/12/2016:
    Manage mandatory properties in java templates.

#### Pogo-9.4.3 - 02/12/2016:
    Remove RcsId in C++ files.

#### Pogo-9.4.2 - 25/11/2016:
    On python templates, fix Copyright on multi lines bug, some pep8 warnings, state and status when overloaded, .....

#### Pogo-9.4.1 - 13/09/2016:
    Improve Python templates.
    Generate CMakeLists.txt only if cmake_tango.opt found.

#### Pogo-9.4.0 - 12/09/2016:
    CMakeLists.txt file could be generated for cmake usage (linux/cpp)..
    Fix a bug in Python for set_change/archive_event. The 2nd parameter was inverted.

#### Pogo-9.3.0 - 04/08/2016:
    Fix bug in property default values in Python code.

#### Pogo-9.2.9 - 12/07/2016:
    Fix problem on resize nodes.
    Check MonoClass/MutiClasses to open chooser.

#### Pogo-9.2.8 - 28/06/2016:
    Manage state machine for dynamic attributes and commands

#### Pogo-9.2.7 - 27/06/2016:
    Template management added.

#### Pogo-9.2.6 - 20/05/2016:
    Fix a problem with event for java dynamic attributes

#### Pogo-9.2.5 - 08/04/2016:
    Add a clone class/device property option menu.
    Add an editor for class description.

#### Pogo-9.2.4 - 07/03/2016:
    Pb in pipe is_allowed method fixed.
    Pb in pipe argument name fixed.

#### Pogo-9.2.3 - 25/02/2016:
    Use of entry_points in setup.py for better compatibility.

#### Pogo-9.2.2 - 04/02/2016:
    Fix a bug in HTML generation with a command description empty

#### Pogo-9.2.1 - 04/02/2016:
    Improve Python and PythonHL templates for inheritance

#### Pogo-9.2.0 - 20/01/2016:
    Improve check name for forwarded attributes.

#### Pogo-9.1.12 - 02/12/2015:
    Fix a problem in relative path for multi classes project.

#### Pogo-9.1.11 - 01/12/2015:
    Bugs in dynamic commands fixed.

#### Pogo-9.1.10 - 25/11/2015:
    Fix bug in DevLong64 java property type
    Improve Python templates (path handling bug fix, protected region for run method)

#### Pogo-9.1.9 - 20/11/2015:
    Fix a bug protected area when rename attribute on c++.
    Fix a problem in relative path for multi classes project.

#### Pogo-9.1.8 - 19/11/2015:
    Add a potected region in main method for java code.
    Add colors on State objects.

#### Pogo-9.1.7 - 02/10/2015:
    Manage relative path for inheritance and multiple classes project.
    Improve State machine window.


pogo6/pogo/appli/TreePopupMenu.java
//+======================================================================
// $Source$
//
// Project:   Pogo
//
// Description:  java source code for the TreePopupMenu class definition .
//
// $Author$
//
// $Revision$
//
// $Log$
// Revision 3.12  2009/02/18 15:03:49  pascal_verdier
// Tango 7 (IDL 4) compatibility implementation.
//
// Revision 3.11  2007/07/30 11:51:43  pascal_verdier
// Bug on popup menu fixed
// (add a set visible and anabled for all items added)
//
// Revision 3.10  2007/04/17 14:03:06  pascal_verdier
// Attributes are now displayed separatly
// (Scalar, Spectrum and image attributes)
//
// Revision 3.9  2006/11/24 14:26:42  pascal_verdier
// Use switch on clicked item.
//
// Revision 3.8  2006/09/22 13:02:16  pascal_verdier
// Bug in code generation directory fixed.
// Edit Abstract class if any added.
//
// Revision 3.7  2006/06/26 09:24:53  pascal_verdier
// ango-5.5 compatiblity.
// extern C method added to be used as shared library.
// VCC 6 project file generated if running on Win32.
// .obj, .so, executable files generated in separated directories.
//
// Revision 3.6  2005/11/24 08:28:45  pascal_verdier
// Analized with intelliJid.
//
// Revision 3.5  2005/06/14 08:48:39  pascal_verdier
// SuperClass is replaced by AbstractClass.
//
// Revision 3.4  2005/01/12 12:48:47  pascal_verdier
// State Machine Dialog added.
//
// Revision 3.3  2004/10/21 06:54:47  pascal_verdier
// Check for non double item definitions
// for add, edit and clone item.
//
// Revision 3.2  2004/09/07 12:02:45  pascal_verdier
// Remove CVS log messages and info from templates.
//
// Revision 3.1  2004/08/26 07:25:35  pascal_verdier
// Attributes are now generated as class.
// Look and field changed.
//
//
// copyleft 1995 by European Synchrotron Radiation Facility, Grenoble, France
//               All Rights Reversed
//-======================================================================

package pogo.appli;

import pogo.gene.PogoUtil;
import pogo.gene.Cmd;
import pogo.gene.PogoDefs;

import javax.swing.*;
import javax.swing.tree.DefaultMutableTreeNode;
import java.awt.event.ActionEvent;
import java.awt.event.MouseEvent;

public class TreePopupMenu extends JPopupMenu  implements PogoDefs, PogoAppliDefs
{
	private PogoAppli	appli;
	private PogoTree	parent;
	private boolean		osIsUnix;
	static private String[]	pMenuLabels = {
									"Edit Item",
									"Edit code",
									"Clone",
									"Delete",
									"Move Up",
									"Move Down",
									"Create new ",
									"Sort items",
									"Edit State Machine",
									"Edit descriptions",
                                    "Edit Class",
                                    "Edit Abstract Class",
									"Class summary",
 									"Edit Device Class ID",
                                   };
	
	static private final int	OFFSET       = 2;		//	Label And separator
	static private final int	EDIT_ITEM    = 0;
	static private final int	EDIT_CODE    = 1;
	static private final int	CLONE_OBJ    = 2;
	static private final int	DELETE_ITEM  = 3;
	static private final int	MOVE_UP      = 4;
	static private final int	MOVE_DOWN    = 5;
	static private final int	NEW_ITEM     = 6;
	static private final int	SORT_ITEMS   = 7;
	static private final int	EDIT_ST_MACH = 8;
	static private final int	EDIT_DESC    = 9;
	static private final int	EDIT_CLASS    = 10;
	static private final int	EDIT_ABSTRACT = 11;
	static private final int	CLASS_SUMMARY = 12;
	static private final int	DEV_CLASS_ID  = 13;


 	//===============================================================
	//===============================================================
	public TreePopupMenu(PogoAppli appli, PogoTree parent)
	{
		super();
		this.appli  = appli;
		this.parent = parent;

		osIsUnix = pogo.gene.PogoUtil.osIsUnix();
		buildBtnPopupMenu();
	}
	//===============================================================
	/**
	 *	Create a Popup menu for host control
	 */
	//===============================================================
	private void buildBtnPopupMenu()
	{
		JLabel	title = new JLabel("Edit Menu :");
        title.setFont(new java.awt.Font("Dialog", 1, 16));
		add(title);
		add(new JPopupMenu.Separator());

		for (int i=0 ; i<pMenuLabels.length ; i++)
		{
			JMenuItem	btn = new JMenuItem(pMenuLabels[i]);
   			btn.addActionListener (new java.awt.event.ActionListener () {
				public void actionPerformed (ActionEvent evt) {
            		itemActionPerformed(evt);
				}
			});
			add(btn);
		}
	}
	//======================================================
	//======================================================
	public void showMenu(MouseEvent evt)
	{
		DefaultMutableTreeNode node = parent.getSelectedNode();
		if (node==null)
			return;

		for (int i=0 ; i<pMenuLabels.length ; i++)
		{
			((JMenuItem)getComponent(OFFSET+i)).setVisible(true);
			((JMenuItem)getComponent(OFFSET+i)).setEnabled(true);
		}
		Object obj = node.getUserObject();
		JMenuItem	new_item     = (JMenuItem)getComponent(OFFSET+NEW_ITEM);
		JMenuItem	sort_items   = (JMenuItem)getComponent(OFFSET+SORT_ITEMS);
		JMenuItem	edit_desc    = (JMenuItem)getComponent(OFFSET+EDIT_DESC);

		JMenuItem	clone_item  = (JMenuItem)getComponent(OFFSET+CLONE_OBJ);
		clone_item.setText("Clone");
        getComponent(OFFSET+EDIT_CLASS).setVisible(false);
        getComponent(OFFSET+EDIT_ABSTRACT).setVisible(false);
        getComponent(OFFSET+CLASS_SUMMARY).setVisible(false);
        getComponent(OFFSET+DEV_CLASS_ID).setVisible(false);

		//	Check witch items are enabled
		switch(PogoUtil.instanceOf(obj))
		{
		case CLASS_ROOT:
            for (int i=0 ; i<pMenuLabels.length ; i++)
                getComponent(OFFSET+i).setVisible(false);
            getComponent(OFFSET+EDIT_CLASS).setVisible(true);
            getComponent(OFFSET+CLASS_SUMMARY).setVisible(true);
            getComponent(OFFSET+DEV_CLASS_ID).setVisible(true);
            if (!parent.server.inheritedClassName().equals(PogoDefs.tangoDeviceImpl))
                getComponent(OFFSET+EDIT_ABSTRACT).setVisible(true);
            break;

        case COMMANDS:
			Cmd	cmd = (Cmd)obj;
			boolean	enabled = !(cmd.virtual_method);
			getComponent(OFFSET+CLONE_OBJ).setEnabled(enabled);
			getComponent(OFFSET+DELETE_ITEM).setEnabled(enabled);
			getComponent(OFFSET+MOVE_UP).setEnabled(enabled);
			getComponent(OFFSET+MOVE_DOWN).setEnabled(enabled);

			enabled = (enabled || cmd.override_method==PogoDefs.NOT_OVERRIDE);
			getComponent(OFFSET+EDIT_ITEM).setEnabled(enabled);

			enabled = (enabled || cmd.override_method==PogoDefs.ALREADY_OVERRIDED);
			getComponent(OFFSET+EDIT_CODE).setEnabled(enabled && osIsUnix);

			new_item.setVisible(false);
			sort_items.setVisible(false);
			edit_desc.setVisible(false);
			break;

		case STATES:
			boolean	edit_state = osIsUnix && (appli.getLanguage()==PogoDefs.cppLang);
			getComponent(OFFSET+EDIT_ITEM).setEnabled(true);
			getComponent(OFFSET+EDIT_CODE).setEnabled(edit_state);
			getComponent(OFFSET+CLONE_OBJ).setEnabled(true);
			getComponent(OFFSET+DELETE_ITEM).setEnabled(true);
			getComponent(OFFSET+MOVE_UP).setEnabled(true);
			getComponent(OFFSET+MOVE_DOWN).setEnabled(true);
			new_item.setVisible(false);
			sort_items.setVisible(false);
			edit_desc.setVisible(false);
			break;

		case ATTRIBUTES:
		case PROPERTY:
			getComponent(OFFSET+EDIT_ITEM).setEnabled(true);
			getComponent(OFFSET+EDIT_CODE).setEnabled(osIsUnix);
			getComponent(OFFSET+CLONE_OBJ).setEnabled(true);
			getComponent(OFFSET+DELETE_ITEM).setEnabled(true);
			getComponent(OFFSET+MOVE_UP).setEnabled(true);
			getComponent(OFFSET+MOVE_DOWN).setEnabled(true);
			new_item.setVisible(false);
			sort_items.setVisible(false);
			edit_desc.setVisible(false);
			
			//	Special management for properties
			if (node.getParent().toString().indexOf("Class")>=0)
				clone_item.setText("Clone as Device Property");
			else
			if (node.getParent().toString().indexOf("Device")>=0)
				clone_item.setText("Clone as Class Property");
			break;

		case STRING_NODE:
			if (node.getParent().getChildCount()>2)
				sort_items.setVisible(true);
			else
				sort_items.setVisible(false);
			new_item.setVisible(true);
			new_item.setText(pMenuLabels[NEW_ITEM] + obj.toString());
			edit_desc.setVisible(true);

			getComponent(OFFSET+EDIT_ITEM).setEnabled(false);
			getComponent(OFFSET+EDIT_CODE).setEnabled(false);
			getComponent(OFFSET+CLONE_OBJ).setEnabled(false);
			getComponent(OFFSET+DELETE_ITEM).setEnabled(false);
			getComponent(OFFSET+MOVE_UP).setEnabled(false);
			getComponent(OFFSET+MOVE_DOWN).setEnabled(false);
			break;
		}
		show(parent, evt.getX(), evt.getY());
	}
	//===============================================================
	//===============================================================
	private void itemActionPerformed(ActionEvent evt)
	{
		DefaultMutableTreeNode node = parent.getSelectedNode();
        Object  obj = evt.getSource();
        int cmdidx = 0;
        for (int i=0 ; i<pMenuLabels.length ; i++)
            if (getComponent(OFFSET+i)==obj)
                cmdidx = i;

        switch(cmdidx)
        {
        case EDIT_ITEM:
			parent.editItem();
		    break;
		case EDIT_CODE:
			parent.editCode();
		    break;
		case CLONE_OBJ:
			parent.cloneLeaf(node);
    		break;
		case DELETE_ITEM:
			parent.removeSelectedLeaf();
            break;
		case MOVE_UP:
			parent.moveSelection("Up");
		    break;
		case MOVE_DOWN:
			parent.moveSelection("Down");
            break;
		case NEW_ITEM:
			parent.createItem();
            break;
		case SORT_ITEMS:
			parent.sortItems(node);
            break;
		case EDIT_ST_MACH:
			parent.editStateMachine(node);
		    break;
		case EDIT_DESC:
			parent.editDescriptions(node);
            break;
		case EDIT_CLASS:
			parent.editClassObject();
            break;
		case EDIT_ABSTRACT:
			parent.editAbstractClass();
            break;
		case CLASS_SUMMARY:
			parent.showClassSummary();
            break;
 		case DEV_CLASS_ID:
			parent.editDeviceClassID();
            break;
       }
    }
}

